### PR TITLE
Events webhook API endpoint

### PR DESCRIPTION
**Pull request by nico-stefani**
|Related issue|
|---|
|#16068|

## Description

This PR adds a new API endpoint with the capability of ingesting events to analysisd.

### CODE REVIEW

File: api/api/configuration.py
Comment (nico-stefani) [inline api/api/configuration.py:88]: This value will be defined based on performance. Is bypassed for that reason.

File: framework/wazuh/rbac/default/policies.yaml
Comment (Selutario) [inline framework/wazuh/rbac/default/policies.yaml:368]:
```suggestion
    description: Allow sending events to analysisd.
```
or
```suggestion
    description: Allow ingesting events.
```

**Review by Selutario — CHANGES_REQUESTED** [review]

File: api/api/configuration.py
Comment (nico-stefani) [reply api/api/configuration.py:88]: Yes, I will take `max_events_per_request`, it is more appropriate.

File: api/api/spec/spec.yaml
Comment (nico-stefani) [reply api/api/spec/spec.yaml]: Ok, I agree, I will do the change. As we talked about, if in the future we add some metadata will be added for the bulk.

File: api/api/middlewares.py
Comment (nico-stefani) [reply api/api/middlewares.py:21]: Yes, we talked about and we choose this behavior in order to keep it simple for the users. So, only will be needed to modify the value of `max_request_per_minute` to change this limit.
Of course, must we write proper documentation explaining the case.
@davidjiglesias what do you think about that?

File: api/api/spec/spec.yaml
Comment (nico-stefani) [reply api/api/spec/spec.yaml]: I removed the format key.

File: api/api/middlewares.py
Comment (davidjiglesias) [reply api/api/middlewares.py:21]: Yes, we decided that but after talking with Selu it is not so clear, I will try to set up a 10 minutes meeting to finalize the details.

File: api/api/middlewares.py
Comment (Selutario) [inline api/api/middlewares.py:154]:
```suggestion
    """Rate limiter for the events ingestion endpoint.

    This function checks that the maximum number of requests per minute passed in `max_requests` is not exceeded.

    Parameters
```

File: framework/wazuh/core/wazuh_queue.py
Comment (Selutario) [inline framework/wazuh/core/wazuh_queue.py:193]:
```suggestion
```

**Review by Selutario — CHANGES_REQUESTED** [review]

**Review by Selutario — CHANGES_REQUESTED** [review]
### API log size
The events are included in the api.log in INFO level, which can significantly increase the size of the API log in some use cases. On the other hand, it is the expected behavior since both the body and the parameters are logged for all endpoints.

Maybe we should discuss whether to make an exception for this endpoint or not, @davidjiglesias @nico-stefani  
```
======== Running on https://0.0.0.0:55000 ========
(Press CTRL+C to quit)
2023/05/16 09:29:19 INFO: wazuh 172.19.0.1 "GET /security/user/authenticate" with parameters {} and body {} done in 0.585s: 200
2023/05/16 09:30:00 INFO: wazuh 172.19.0.1 "POST /events" with parameters {"wait_for_complete": "true"} and body {"events": ["May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A"]} done in 5.490s: 200
```

### Logs order
In my tests, when there are multiple logs in the same request, even though the API is sending them to analysisd in the correct order (the one set in the list inside the body), analysisd is generating the alerts in random order. It's not the API's fault, but I don't know if it could be a problem.

**Comment by davidjiglesias** [comment]
> ### API log size
> The events are included in the api.log in INFO level, which can significantly increase the size of the API log in some use cases. On the other hand, it is the expected behavior since both the body and the parameters are logged for all endpoints.
> 
> Maybe we should discuss whether to make an exception for this endpoint or not, @davidjiglesias @nico-stefani
> 
> ```
> ======== Running on https://0.0.0.0:55000 ========
> (Press CTRL+C to quit)
> 2023/05/16 09:29:19 INFO: wazuh 172.19.0.1 "GET /security/user/authenticate" with parameters {} and body {} done in 0.585s: 200
> 2023/05/16 09:30:00 INFO: wazuh 172.19.0.1 "POST /events" with parameters {"wait_for_complete": "true"} and body {"events": ["May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A", "May 16 10:20:48 selu-pc slack.desktop[4245]: [05/16/23, 10:20:48:248] info: [DND_V2] (TMYFNFQ3W) Checking for changes in DND status for the following members: UQS527Y4E,UPBJ806RJ,U04EK45FSD8,U04GCHUJD43,U04RLAR94NS,U0503TGR3KP,U04RQ0ZBZNZ,U03P70SNNU9,U03N2TB2VB3,U0509ABU5LL,U04RSUHKPD0,U051V99UWGH,UPC2M2CRK,UQZH2L3K3,UUFSGP6UW,U01Q59P9CA2,U02KLL25TB2,U04DNBPATCH,U02TDBK9B4L,U02768GCBL7,U03GADDM4QH,UP8B51YRG,U024SGQ5PL0,UPC2M6ELV,U01LES5UYJU,U018RJ2U5FY,U026YHMBB44,U01NABY0K0A"]} done in 5.490s: 200
> ```
> 
> ### Logs order
> In my tests, when there are multiple logs in the same request, even though the API is sending them to analysisd in the correct order (the one set in the list inside the body), analysisd is generating the alerts in random order. It's not the API's fault, but I don't know if it could be a problem.

I think we should only show the events in DEBUG mode, otherwise, they have the potential to overflow the log easily.

Regarding the alerts order, I believe analysisd is mutl-threaded and that is why this can happen. I do not think this is in our realm nor a big issue. This development is not intended for fast-paced injection.

File: api/api/alogging.py
Comment (Selutario) [inline api/api/alogging.py:1]: This seems not to be working when using a token obtained by `Authorization context` instead of by user/password. As it can be seen below, all events are included even when API is in `INFO` level:
```
root@wazuh-master:/# /var/ossec/bin/wazuh-apid -f
Starting API in foreground
2023/05/18 08:29:54 INFO: Checking RBAC database integrity...
2023/05/18 08:29:54 INFO: /var/ossec/api/configuration/security/rbac.db file was detected
2023/05/18 08:29:54 INFO: RBAC database integrity check finished successfully
2023/05/18 08:29:59 INFO: Listening on 0.0.0.0:55000..
======== Running on https://0.0.0.0:55000 ========
(Press CTRL+C to quit)
2023/05/18 08:30:01 INFO: wazuh-wui 172.21.0.1 "GET /security/user/authenticate" with parameters {} and body {} done in 0.606s: 200
2023/05/18 08:30:02 INFO: wazuh-wui (7e3bd7c5ea0e07beb34e9dbef70eae6d) 172.21.0.1 "POST /events" with parameters {"wait_for_complete": "true"} and body {"events": ["wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7", "wzh-inventory Hostname: XXXXX IPV4: 0.0.0.0 Serial:  VMware-42 04  Make:  VMware, Inc. Model:  VMware Virtual Platform OS: CentOS Linux release 7.9.2009 (Core) Kernel: 3.10.0-1160.76.1.el7.x86_64 Function: Production Product: PROD Type: APP CPU: 4 Memory: 7"]} done in 0.121s: 200
```

In my opinion, the `custom_logging` function was already repeating code unnecessarily and is getting a bit more complicated than necessary. I think it's a good time to refactor it, maybe something like this:
```PYTHON
    def custom_logging(self, user: str, remote: str, method: str, path: str, query: dict, body: dict, time: float,
                       status: int, hash_auth_context: str = ''):
        """Provide the log entry structure depending on the logging format.

        Parameters
        ----------
        user : str
            User who perform the request.
        remote : str
            IP address of the request.
        method : str
            HTTP method used in the request.
        path : str
            Endpoint used in the request.
        query : dict
            Dictionary with the request parameters.
        body : dict
            Dictionary with the request body.
        time : float
            Required time to compute the request.
        status : int
            Status code of the request.
        hash_auth_context : str, optional
            Hash representing the authorization context. Default: ''
        """
        json_info = {
            'user': user,
            'ip': remote,
            'http_method': method,
            'uri': f'{method} {path}',
            'parameters': query,
            'body': body,
            'time': f'{time:.3f}s',
            'status_code': status
        }

        if not hash_auth_context:
            log_info = f'{user} {remote} "{method} {path}" '
        else:
            log_info = f'{user} ({hash_auth_context}) {remote} "{method} {path}" '
            json_info['hash_auth_context'] = hash_auth_context

        if path != '/events':
            log_info += f'with parameters {json.dumps(query)} and body {json.dumps(body)} done in {time:.3f}s: {status}'
        elif self.logger.level >= 20:
            # If log level is info simplify the messages and remove the body value.
            events = body.get('events', [])
            json_info.pop('body')
            json_info['number_of_events'] = len(events)
            log_info += f'with parameters {json.dumps(query)} and {len(events)} events done in {time:.3f}s: {status}'

        self.logger.info(log_info, extra={'log_type': 'log'})
        self.logger.info(json_info, extra={'log_type': 'json'})
```

**Review by Selutario — CHANGES_REQUESTED** [review]

**Review by Selutario — APPROVED** [review]
LGTM

File: framework/wazuh/event.py
Comment (nico-stefani) [reply framework/wazuh/event.py:24]: @davidjiglesias I was working on the validation of the event size, but I have some doubts:

* The size should include the message header? (msg_header + msg) < 64Kb?
* If one event exceeds the limit which should be the behavior?  Return the `AffectedItemsWazuhResult` response maybe?

Comment (Selutario) [reply framework/wazuh/event.py:24]: Hi Nico, I can help with those questions:

> The size should include the message header? (msg_header + msg) < 64Kb?

Yes, 64KB includes the header size and the message. However, the size is not always 64KB, it could be changed by the user or in future releases. Maybe it should be handled in a similar way to wodles like gcloud or aws:
https://github.com/wazuh/wazuh/blob/e6819b2c2f37a9d5fe45873dcd9f6d1a24259445/wodles/aws/aws_s3.py#L356-L372

> If one event exceeds the limit which should be the behavior? Return the AffectedItemsWazuhResult response maybe?

Yes, returning AffectedItemsWazuhResult with the number of failed logs is a good option. Keep in mind that the endpoint must always answer with the same response format, so use `AffectedItemsWazuhResult` for both successful and failed responses.

File: api/api/configuration.py
Comment (davidjiglesias) [inline api/api/configuration.py:87]: ?

File: api/api/configuration.py
Comment (davidjiglesias) [inline api/api/configuration.py:108]: Please try to avoid these kind of changes, it creates issues and conflicts that are unnecessary.

File: api/api/middlewares.py
Comment (davidjiglesias) [inline api/api/middlewares.py:19]: The default max requests is already defined somewhere else (default config), why do we need to declare another variable?

File: api/api/spec/spec.yaml
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:19079]: We need to ensure that the 100 bulk size limit and the 30 max POST events EPS are explained somewhere in the documentation!

File: api/scripts/wazuh-apid.py
Comment (davidjiglesias) [inline api/scripts/wazuh-apid.py:34]: Beware that adding a new process can affect many different places, it is very important to test this in deployments, etc!

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

File: api/scripts/wazuh-apid.py
Comment (nico-stefani) [reply api/scripts/wazuh-apid.py:34]: How do you suggest handling this? Do we need to open a new issue to track it?

Comment (Selutario) [reply api/scripts/wazuh-apid.py:34]: Yes @nico-stefani, you can open a new issue inside the epic if you prefer. As long as this is tested before merging the epic branch, is fine.

File: framework/wazuh/core/wazuh_queue.py
Comment (Selutario) [inline framework/wazuh/core/wazuh_queue.py:82]: I've tested this and, despite what I said [here](https://github.com/wazuh/wazuh/pull/17023#discussion_r1218187732), it doesn't seem to be working right. For example, I am sending an event consisting of a list of numbers: "1 2 3 4 5 ... 26849 26850 26851". This is about 150000 bytes, but no socket.error is being thrown (although it does raise when sizes get even larger).

However in my tests, I can see that even though no error is produced, the last number I find in the archives.log is `12771` (about 65522 bytes). Therefore, part of the event has been discarded without warning the user, so this method is not effective for what we need (we should also open an issue to fix it in other integrations like AWS).

Please research how it is done in other parts of Wazuh or ask the core team if necessary about the most effective way to deal with this.

**Review by Selutario — CHANGES_REQUESTED** [review]

File: api/api/middlewares.py
Comment (Selutario) [inline api/api/middlewares.py:143]:
```suggestion
            'extra_message': f'For POST /events endpoint the limit is set to {max_requests} requests.'
```

File: framework/wazuh/core/wazuh_queue.py
Comment (Selutario) [inline framework/wazuh/core/wazuh_queue.py:223]: Max size is bytes (`B`), not bits (`b`). I think these message is a bit easier to understand:
```suggestion
                "The event is too large to be sent to analysisd (maximum is {self.MAX_MSG_SIZE}B)"
```

**Review by Selutario — CHANGES_REQUESTED** [review]

**Review by Selutario — APPROVED** [review]
LGTM, checks are failing in `4.5.0`, `4.6.0` and `master` but this branch is merged in an epic branch.

**Pull request merged**
